### PR TITLE
improve performance and reduce memory consumption of get_product

### LINKED ISSUE #33944: Memory leak & bad performance with WC_Product get_product
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

When trying to avoid the use of globals, we use `wc_get_product()` (like `get_post()` in WP core).

This itself will then `\WC_Product_Factory->get_product()` which will **always** construct and return a new class instance for that product.
When using `global $product` this isn't an issue, however when replacing it with `wc_get_product()` calls, it quickly gets an issue. A normal WooCommerce shop with multiple templates in a catalog loop now has 1000s of unnecessary duplicate WC_Product objects per page increasing memory usage unnecessarily.

### Expected behavior

Reuse the product from global

### Actual behavior

Creating new instance of product everytime

### Steps to reproduce

call wc_get_product()

### WordPress Environment

WP 6.0
WooCommerce 6.7.0

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by kkmuffme**
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

Closes https://github.com/woocommerce/woocommerce/issues/33944

### How to test the changes in this Pull Request:

call wc_get_product() in the loop multiple times

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [x] Have you written new tests for your changes, as applicable?
-   [] Have you successfully run tests with your changes locally?
-   [] Have you created a changelog file for each project being changed, ie `pnpm changelog add --filter=<project>`?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

**Comment by kkmuffme** [comment]
Are the tests working? Things that are completely unrelated to the changed code are failing??

Please advice on/add the changelog as you see fit, as I don't want to redo that 20 times if you're not happy with the text.

**Comment by barryhughes** [comment]
I understand the intent, but I'm not sure we can accept this as it would introduce:

- Inconsistencies between different versions of WooCommerce.
- Inconsistencies depending on whether or not the fetched product happens have the same ID as the global product object.

These things could introduce subtle and hard-to-troubleshoot bugs in lots of existing code (and indeed new code). To see what I mean, consider the following:

```php
function plugin_a_callback( $product_id ) {
	// Callback starts updating product properties.
	$product = wc_get_product( $product_id );
	$product->set_stock_quantity( $product->get_stock_quantity() + 1 );

	// For some reason, this callback decides to aborts its update of the product.
	if ( false ) {
		$product->save();
	}
}

function plugin_b_callback( $product_id ) {
	// Callback updates product properties and commits the changes.
	$product = wc_get_product( $product_id );
	$product->set_stock_quantity( $product->get_stock_quantity() + 1 );
	$product->save();
}

function test_changes( $product_id, $make_global ) {
	// Reset stock levels for consistency.
	$product = wc_get_product( $product_id );
	$product->set_stock_quantity( 100 );
	$product->save();

	// Maybe make the product global.
	if ( $make_global ) {
		$GLOBALS['product'] = $product;
	}

	// Simulate calling two different plugins that both act on the same product.
	plugin_a_callback( $product_id );
	plugin_b_callback( $product_id );

	// Results...
	var_dump( $product->get_stock_quantity() );
	var_dump( wc_get_product( $product_id )->get_stock_quantity() );
}

$product_id = 12345;                // Set to a valid ID.
test_changes( $product_id, false ); // First time: do *not* set or operate on the global $product
test_changes( $product_id, true );  // Second time: we *do* act on the global $product
```

Basically, it becomes difficult to know if we are dealing with the exact same instance or fresh instances (that just happen to represent the same underlying product record).

**Comment by kkmuffme** [comment]
Thanks for the feedback. WP core uses a similar logic in `get_post` which is the WP core equivalent to `get_product`
See https://developer.wordpress.org/reference/functions/get_post/#source

Therefore, to keep things consistent, WC get_product should behave like WP get_post. I can adjust the PR to make the logical (almost) identical to WP `get_post` if that's preferred

**Comment by barryhughes** [comment]
> Therefore, to keep things consistent, WC get_product should behave like WP get_post.

It generally does—but perhaps we are misunderstanding one another, or else we are thinking about different outcomes. If I can share another snippet:

```php
( function () {
	// Select a valid post and product to use in this test.
	$post_id    = 123;
	$product_id = 456;

	// When fetching a post or product by ID, both get_posts() and
	// wc_get_product() return distinct objects.
	$post_1    = get_post( $post_id );
	$post_2    = get_post( $post_id );
	$product_1 = wc_get_product( $product_id );
	$product_2 = wc_get_product( $product_id );
	var_dump( ( $post_1 !== $post_2 ) );       // => (bool) true
	var_dump( ( $product_1 !== $product_2 ) ); // => (bool) true

	// Let's set up our global post and global product objects.
	$GLOBALS['post']    = get_post( $post_id );
	$GLOBALS['product'] = wc_get_product( $product_id );

	// If we repeat the first test and specify the post/product
	// by ID (important), we again receive distinct objects.
	$post_1    = get_post( $post_id );
	$post_2    = get_post( $post_id );
	$product_1 = wc_get_product( $product_id );
	$product_2 = wc_get_product( $product_id );
	var_dump( ( $post_1 !== $post_2 ) );       // => (bool) true
	var_dump( ( $product_1 !== $product_2 ) ); // => (bool) true

	// Only if we omit the ID do we get back identical objects.
	// This is true of both functions.
	$post_1    = get_post();
	$post_2    = get_post();
	$product_1 = wc_get_product();
	$product_2 = wc_get_product();
	var_dump( ( $post_1 !== $post_2 ) );       // => (bool) false
	var_dump( ( $product_1 !== $product_2 ) ); // => (bool) false
} )();
```

The output should be:

```
bool(true)
bool(true)
bool(true)
bool(true)
bool(false)
bool(false)
```

That is, it is only in the final case where we ask `get_post()` or `wc_get_product()` for the current global object (by not specifying an ID) where we actually get back the same object ... and both functions already behave the same way in this regard.

If I'm understanding the change you proposed here, we would see the opposite result for `wc_get_product()` in the second test (where we specify the ID, which happens to match the global product ID)—which is not aligned with how `get_post()` works.

**Comment by barryhughes** [comment]
I'll go ahead and close this, but you can let us know if you want to re-open (or indeed a fresh PR can easily be created). Thanks!
